File: Data Structures & Algorithms/max-water-container/submission-2.cpp
class Solution {
public:
    int maxArea(vector<int>& heights) {
        /*int result = 0;

        for (int i = 0; i < heights.size(); ++i) {
            for (int j = i + 1; j < heights.size(); ++j) {
                result = max(result, min(heights[i], heights[j]) * (j - i));
            }
        }

        return result;*/

        int result = 0;
        int l = 0;
        int r = heights.size() - 1;

        while (l < r) {
            int area = (r - l) * min(heights[l], heights[r]);
            result = max(result, area);

            if (heights[l] <= heights[r]) {
                l++;
            } else {
                r--;
            }
        }

        return result;
    }
};

File: Data Structures & Algorithms/path-with-minimum-effort/submission-2.py
class Solution:
    def minimumEffortPath(self, heights: List[List[int]]) -> int:
        ROWS, COLS = len(heights), len(heights[0])
        min_heap = [[0, 0, 0]] # [diff, row, col]
        visited = set()
        while min_heap:
            diff, r, c = heapq.heappop(min_heap)
            if (r, c) in visited: continue
            visited.add((r, c))
            if (r, c) == (ROWS - 1, COLS - 1): return diff

            # go through every neighbor
            for dr, dc in [(1, 0), (-1, 0), (0, 1), (0, -1)]:
                nr, nc = r + dr, c + dc
                if (nr < 0 or nr >= ROWS or nc < 0 
                    or nc >= COLS or (nr, nc) in visited): continue
                new_diff = max(diff, abs(heights[r][c] - heights[nr][nc]))
                heapq.heappush(min_heap, [new_diff , nr, nc])
        return 0